### PR TITLE
reef: osd: full-object read CRC mismatch due to 'truncate' modifying oi.size w/o clearing 'data_digest'

### PR DESCRIPTION
**Pull request by NitzanMordhai**
backport tracker: https://tracker.ceph.com/issues/66137

---

backport of https://github.com/ceph/ceph/pull/55008
parent tracker: https://tracker.ceph.com/issues/53240

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/main/src/script/ceph-backport.sh

### CODE REVIEW

**Review by Matan-B — CHANGES_REQUESTED** [review]
https://github.com/ceph/ceph/pull/57616

File: src/test/pybind/test_rados.py [hunk 1]
```diff
@@ -312,8 +312,12 @@ def test_cmpext(self):
     def test_list_objects_empty(self):
         eq(list(self.ioctx.list_objects()), [])
 
-    def test_list_objects(self):
+    def test_read_crc(self):
         self.ioctx.write('a', b'')
+        self.ioctx.write('a', b'', 5)
+        self.ioctx.read('a')
+
+    def test_list_objects(self):
```
Comment (rzarzynski) [inline src/test/pybind/test_rados.py:320]: Yeah, let's backport the commit from https://github.com/ceph/ceph/pull/57616 altogether.

**Review by rzarzynski — CHANGES_REQUESTED** [review]

**Comment by NitzanMordhai** [comment]
all the PRs merged, can we continue?

**Comment by rzarzynski** [comment]
@NitzanMordhai: we need the reef backport. I can't find a link in https://tracker.ceph.com/issues/66531.

**Comment by rzarzynski** [comment]
@NitzanMordhai: how about cherry-picking the commits for https://tracker.ceph.com/issues/66226 in this PR?

**Comment by NitzanMordhai** [comment]
@rzarzynski cherry picked 4 commits into that one.

File: src/test/librados/io_cxx.cc [hunk 2]
```diff
@@ -846,7 +846,7 @@ TEST_F(LibRadosIoECPP, RmXattrPP) {
 
 TEST_F(LibRadosIoECPP, CrcZeroWrite) {
   SKIP_IF_CRIMSON();
-  set_allow_ec_overwrites(pool_name, true);
+  set_allow_ec_overwrites();
```
Comment (rzarzynski) [inline src/test/librados/io_cxx.cc:849]: The new commit lacks the `cherry-picked from ` information (`-x` for `git cherry-pick`).

Comment (NitzanMordhai) [reply src/test/librados/io_cxx.cc:849]: fixed that

**Review by rzarzynski — APPROVED** [review]

**Comment by Matan-B** [comment]
rados approved
https://tracker.ceph.com/projects/rados/wiki/REEF#httpstrackercephcomissues68371

**Comment by yuriw** [comment]
@Matan-B pls remove block

**Pull request merged**